DOC: polish QAP docs for consistency and readability

pmla opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### What does this implement/fix?
Here I have made some aesthetic changes to the `scipy.optimize.quadratic_assignment` module. Mostly it is just changes to the docs.

@mdhaber

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -589,14 +583,15 @@ def _doubly_stochastic(P, tol=1e-3):
 
     This function solves the Quadratic Assignment Problem (QAP) and the
-    Graph Matching Problem (GMP) using the 2-opt algorithm [3]_.
+    Graph Matching Problem (GMP) using the 2-opt algorithm [1]_.
```
The intent was to match the reference numbers across the different methods.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -50,21 +50,21 @@ def quadratic_assignment(A, B, method="faq", options=None):
         A dictionary of solver options. All solvers support the following:
 
-        partial_match : 2d-array of integers, optional, (default = None)
+        maximize : bool (default: False)
```
I agree there should not be a comma before the parentheses. Do you know if a standard for specifying defaults like this (comma vs colon) is written anywhere? It's not in the [Guide to NumPy/SciPy Documentation](https://numpy.org/doc/stable/docs/howto_document.html). This is fine; I'm just wondering if you know of a standard.

pmla replied:
I don't know of a standard. I grep the codebase and look for the most common pattern. In this case there is no pre-existing `optional, (...`.
On the other hand, there is only a single instance of `optional (...`

mdhaber replied:
Yeah, there's no question about the comma; it shouldn't have one. Oops, I mean to ask:
> Do you know if a standard for specifying defaults like this (**equals sign** vs colon) is written anywhere?

pmla replied:
I don't know of a standard for that either.

Grep shows that there are 11 modules which use colon. Only `scipy/optimize/_linprog_ip.py` uses an equals sign.

mdhaber replied:
Oops, my bad in `_linprog_ip.py`. I guess colon is the de facto standard.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -308,48 +317,36 @@ def _quadratic_assignment_faq(A, B,
-        a :math:`m' \times 1` array of ones, is used. This is the "barycenter"
-        of the search space of doubly-stochastic matrices.
+        doubly stochastic matrix" :math:`J = \mathbf{1} \cdot \mathbf{1}^T/m'`,
+        where :math:`1` is a :math:`m' \times 1` array of ones, is used. This
```
```suggestion
        where :math:`\mathbf{1}` is a :math:`m' \times 1` array of ones, is used. This
```
Not sure if this will put you over the line limit.

pmla replied:
I went with a different approach: `This is a :math: m' x m' matrix with all entries equal to :math:1 / m'`.`

This is from [a definition](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.214.3184&rep=rep1&type=pdf):
`The barycenter of B_n is the van der Waerden constant matrix B_∗ with all entries equal to 1/n`

mdhaber replied:
Sounds good.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -50,21 +50,21 @@ def quadratic_assignment(A, B, method="faq", options=None):
-            Setting `maximize` to ``True`` solves the Graph Matching Problem
-            (GMP) rather than the Quadratic Assingnment Problem (QAP).
-
         rng : {None, int, `~np.random.RandomState`, `~np.random.Generator`}
```
```suggestion
        rng : {None, int, `~np.random.RandomState`, `~np.random.Generator`} (default: None)
```
Might as well specify the default in this line since we're doing that for all the other options. You could also remove it from the end of the parameter description.

mdhaber replied:
Also, since we've been at this level of nit for a while:
![image](https://user-images.githubusercontent.com/6570539/93125631-4a3d4e80-f680-11ea-98e8-37fdafafd6f1.png)
Sometimes RandomState has single backticks, sometimes double. My understanding is that options/parameters standing alone should be single backtick, but code involving them should be double. 

Not that there's any hope of consistency, given that our sphinx theme renders the parameters in bold - not italicized OR teletype - in the headings of the Parameters section. Personally I think they should _always_ be teletype.

pmla replied:
We were missing a default on the first line here. I couldn't fit it all in 80 chars whilst including the default. Instead, I adapted the description from [`spatial.transform.Rotation.random'](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.random.html#scipy.spatial.transform.Rotation.random) which is very similar to our needs here and very concise.

mdhaber reviewed (commented):
Also in the notes:
![image](https://user-images.githubusercontent.com/6570539/93126588-6392ca80-f681-11ea-9417-018f713fa24c.png)
Should be "it typically offers"

sethtroisi reviewed (approved):
These are all improvements in my eye.

pmla commented:
@mdhaber Thanks for the review.

I have made another pass. In addition to your suggestions I have tried to make some of the descriptions more concise and fixed some minor errors.

* The `P0` parameter had an incomplete definition of doubly-stochastic matrices (non-negativity is also required). Rather than write out the full definition here I have added a link to the wikipedia page.
* I have rewritten the definition of the `shuffle_input` parameter. In my opinion the current description does not say what it is doing, nor when it has an effect. This led to a lot of back and forth in #12775.
* The 'tol' parameter is not actually measuring the "relative Frobenius norm". I have removed reference to this.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -636,55 +606,44 @@ def _quadratic_assignment_2opt(A, B, maximize=False, partial_match=None,
+        global `numpy.random` random state.
+
+    partial_match : 2-D array of integers, optional (default: None)
+        Fixes part of the matching. Also known as a "seed" [2]_.
```
['scipy.optimize.quadratic_assignment', ':impl: scipy.optimize._qap._quadratic_assignment_2opt', ':method: 2opt']:71: WARNING: Unknown target name: "2".
/home/circleci/repo/doc/source/optimize.qap-faq.rst:92: WARNING: Inline interpreted text or phrase reference start-string without end-string.

Need to add [2] to the references of this function.

pmla replied:
Thanks for debugging that. I was getting confused with `[2]` and `2opt` - thought it was complaining about the latter.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -50,54 +50,44 @@ def quadratic_assignment(A, B, method="faq", options=None):
+            Each row of `partial_match` specifies a pair of matched nodes:
+            node ``partial_match[i, 0]`` of `A` is matched to node
+            ``partial_match[i, 1]`` of `B`. The array has shape ``(m, 2)``,
+            where ``m`` is not greater than the number of nodes, :math:`n`.
```
having :math:`n` right next to `.` might be causing the error:
`/home/circleci/repo/doc/source/optimize.qap-faq.rst:92: WARNING: Inline interpreted text or phrase reference start-string without end-string.`

![image](https://user-images.githubusercontent.com/6570539/93405842-fa5ab500-f842-11ea-9628-1aac53c95d5d.png)

@tylerjereddy I'm sorry if you're not the right person to ask, but do you know if we can get the .rst files uploaded as artifacts in CircleCI? It would be a lot easier to debug from line numbers.

tylerjereddy replied:
Presumably might be done with something similar to what we have for html:

```yaml
 57       - store_artifacts:
 58           path: doc/build/html-scipyorg
 59           destination: html-scipyorg
```

https://circleci.com/docs/2.0/artifacts/ 

If you mean taking `.rst` files from around the code base paths and just uploading them, could probably use some shell command or globbing to move them to a single path for upload maybe? I probably don't know enough about the debug scenario to understand the advantage as well as you I guess.

mdhaber replied:
Yes, I think something like that. The reason is that the documentation build errors reference line numbers, but those line numbers correspond with .rst files that we can't see. If we could find the lines, we'd have more context to figure out what to change in the docstrings.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -292,91 +281,71 @@ def _quadratic_assignment_faq(A, B,
+        :math:`K` is a random doubly stochastic matrix.
+
+        If the initial position is an array, it must be a doubly stochastic
+        matrix of size :math:`m' x m'` where :math:`m' = n - m`.
```
```suggestion
        matrix of size :math:`m' \times m'` where :math:`m' = n - m`.
```
Here and everywhere.

pmla replied:
Thanks. Fixed.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -292,91 +281,71 @@ def _quadratic_assignment_faq(A, B,
+        If ``"barycenter"`` (default), the initial position is the barycenter
+        of the Birkhoff polytope (the space of doubly stochastic matrices).
+        This is a :math:`m' x m'` matrix with all entries equal to
+        :math:1 / m'`.
```
```suggestion
        :math:`1 / m'`.
```
Here's the source of the error:
`/home/circleci/repo/doc/source/optimize.qap-faq.rst:92: WARNING: Inline interpreted text or phrase reference start-string without end-string.`

pmla replied:
Thanks for finding that. Fixed.

pmla commented on scipy/optimize/_qap.py:
```diff
@@ -292,91 +281,71 @@ def _quadratic_assignment_faq(A, B,
+        global `numpy.random` random state.
+
+    P0 : "barycenter", "randomized", or a 2-D array (default: "barycenter")
+        Initial guess. Must be a doubly-stochastic matrix [3]_.
```
They are now all "initial position".

pmla commented on scipy/optimize/_qap.py:
```diff
@@ -292,91 +281,71 @@ def _quadratic_assignment_faq(A, B,
+        :math:`K` is a random doubly stochastic matrix.
+
+        If the initial position is an array, it must be a doubly stochastic
+        matrix of size :math:`m' x m'` where :math:`m' = n - m`.
```
I have moved it to the top.

pmla commented on scipy/optimize/_qap.py:
```diff
@@ -292,91 +281,71 @@ def _quadratic_assignment_faq(A, B,
+        If ``"barycenter"`` (default), the initial position is the barycenter
+        of the Birkhoff polytope (the space of doubly stochastic matrices).
+        This is a :math:`m' x m'` matrix with all entries equal to
+        :math:1 / m'`.
```
Fixed by moving P0 to the top.

pmla commented:
Please don't merge this yet.

pmla commented:
I have simplified the description of the maximize parameter.

Graph matching implies `maximize=True`, but the converse is not true. For example, to partition a graph of 9 nodes into three components, maximize=True, where `A` is the adjacency matrix (optionally weighted) and
```
B = [[1, 1, 1, 0, 0, 0, 0, 0, 0],
     [1, 1, 1, 0, 0, 0, 0, 0, 0],
     [1, 1, 1, 0, 0, 0, 0, 0, 0],
     [0, 0, 0, 1, 1, 1, 0, 0, 0],
     [0, 0, 0, 1, 1, 1, 0, 0, 0],
     [0, 0, 0, 1, 1, 1, 0, 0, 0],
     [0, 0, 0, 0, 0, 0, 1, 1, 1],
     [0, 0, 0, 0, 0, 0, 1, 1, 1],
     [0, 0, 0, 0, 0, 0, 1, 1, 1]]
```

rgommers commented:
This looks about ready, it just accumulated a merge conflict. Could you update @pmla?

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -423,6 +385,9 @@ def _quadratic_assignment_faq(A, B,
 
     .. [2] D. Fishkind, S. Adali, H. Patsolic, L. Meng, D. Singh, V. Lyzinski,
            C. Priebe, "Seeded graph matching", Pattern Recognit. 87 (2019):
+           203-215, https://doi.org/10.1016/j.patcog.2018.09.014
+
+    .. [3] https://en.wikipedia.org/wiki/Doubly_stochastic_matrix
            203-215, :doi:`10.1016/j.patcog.2018.09.014`
```
```suggestion
           203-215, :doi:`10.1016/j.patcog.2018.09.014`

    .. [3] https://en.wikipedia.org/wiki/Doubly_stochastic_matrix
```

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -424,6 +386,8 @@ def _quadratic_assignment_faq(A, B,
            C. Priebe, "Seeded graph matching", Pattern Recognit. 87 (2019):
            203-215, :doi:`10.1016/j.patcog.2018.09.014`
+
+    .. [3] https://en.wikipedia.org/wiki/Doubly_stochastic_matrix
```
```suggestion
    .. [3] “Doubly stochastic Matrix,” Wikipedia.
            https://en.wikipedia.org/wiki/Doubly_stochastic_matrix
```

pmla commented:
Apologies for the delay. I think I'm all done here.

Thanks for the assistance @mdhaber

mdhaber commented:
@asaadeldin11 would you take a look at this?

asaadeldin11 commented:
These changes look good to me

rgommers commented:
All green and everyone is happy here, let's get this in. Thanks @pmla, all!
